Querier crash: invalid UTF-8 in match[]/query panics the active request tracker (index out of range [-1])

**AI Tool Usage Notice**
If you used an AI tool to help draft this issue,
please make sure you have reviewed and validated all content before submitting.
You are responsible for the accuracy and quality of everything in this report.
Low-quality or unreviewed AI-generated submissions may be closed without further investigation.
See our [Generative AI Contribution Policy](https://github.com/cortexproject/cortex/blob/master/GENAI_POLICY.md) for details.

**Describe the bug**
`trimStringByBytes` in the active-request tracker scans backwards for a UTF-8 rune boundary with no lower bound ([`pkg/util/request_tracker/request_extractor.go:82-93`](https://github.com/cortexproject/cortex/blob/master/pkg/util/request_tracker/request_extractor.go#L82-L93)):

```go
func trimStringByBytes(str string, size int) string {
	bytesStr := []byte(str)
	trimIndex := len(bytesStr)
	if size < len(bytesStr) {
		for !utf8.RuneStart(bytesStr[size]) {
			size--
		}
		trimIndex = size
	}
	return string(bytesStr[:trimIndex])
}
```

If the string consists of UTF-8 continuation bytes (`0x80`–`0xBF`), no byte is ever a rune start, `size` underflows past zero and `bytesStr[-1]` panics with `index out of range [-1]`.

This is reachable from an unauthenticated-by-content, tenant-supplied request parameter, and the panic happens on a goroutine with no `recover()`, so it terminates the querier process:

1. The active request tracker is **enabled by default** — `-querier.active-query-tracker-dir` defaults to `./active-query-tracker` ([`pkg/querier/querier.go:152`](https://github.com/cortexproject/cortex/blob/master/pkg/querier/querier.go#L152)) — and `ApiExtractor` wraps the whole Prometheus API router, including `/api/v1/series`, `/api/v1/labels` and `/api/v1/label/{name}/values` ([`pkg/api/handlers.go:307-316`](https://github.com/cortexproject/cortex/blob/master/pkg/api/handlers.go#L307-L316), route registered at [`pkg/api/api.go:467`](https://github.com/cortexproject/cortex/blob/master/pkg/api/api.go#L467)).
2. The query-frontend PromQL-parses only the `query`/`query_range` operations ([`pkg/querier/tripperware/roundtrip.go:195-200`](https://github.com/cortexproject/cortex/blob/master/pkg/querier/tripperware/roundtrip.go#L195-L200)), so a `series` request's `match[]` value is forwarded to the querier byte-for-byte with no validation.
3. On the querier, the frontend/scheduler worker runs each request in a bare `go func()` with no `recover()` ([`pkg/querier/worker/scheduler_processor.go:141`](https://github.com/cortexproject/cortex/blob/master/pkg/querier/worker/scheduler_processor.go#L141)), and `httpgrpc_server.Server.Handle` invokes the router via `s.handler.ServeHTTP(recorder, req)` directly rather than through `net/http`'s server, so `net/http`'s per-connection panic recovery does not apply. I checked `vendor/github.com/weaveworks/common/` and found no recovery middleware in this chain. An unrecovered panic in a goroutine terminates the process.

Note this is the same class of panic that #7640 set out to fix; that PR added a `repeatSize <= 0` guard in `trimForJsonMarshalRecursive` but did not touch the underflow inside `trimStringByBytes`. Its regression tests use only valid multi-byte UTF-8 (`世`), which always terminates the backwards scan at index 0, so the failing case stays uncovered.

**To Reproduce**
Steps to reproduce the behavior:
1. Start Cortex master (85a4553) with default settings, queriers connected to a query-frontend or query-scheduler.
2. As any authenticated tenant, send a request whose `match[]` value is long enough to exceed `maxEntrySize` (1000) and consists of UTF-8 continuation bytes:
   `GET /prometheus/api/v1/series?match%5B%5D=<1200 repetitions of %80>`
3. The querier process panics and exits. Repeating the request keeps queriers crash-looping.

Unit-level reproduction (I verified all three of these on 85a4553 — each panics at `request_extractor.go:86`):
```go
// via the API extractor, mimicking /api/v1/series
req := httptest.NewRequest("GET", "/api/v1/series?match%5B%5D="+strings.Repeat("\x80", 1200), nil)
(&ApiExtractor{}).Extract(req)

// via the instant-query extractor
form := "query=" + strings.Repeat("\x80", 1200)
req := httptest.NewRequest("POST", "/api/v1/query", strings.NewReader(form))
req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
(&InstantQueryExtractor{}).Extract(req)

// directly
trimForJsonMarshal(strings.Repeat("\x80", 1200), 900)
```

<details>
<summary>Panic output</summary>

```
--- FAIL: TestZZVerifyReproInvalidUTF8Panic (0.00s)
panic: runtime error: index out of range [-1] [recovered, repanicked]

github.com/cortexproject/cortex/pkg/util/request_tracker.trimStringByBytes(...)
	pkg/util/request_tracker/request_extractor.go:86
github.com/cortexproject/cortex/pkg/util/request_tracker.trimForJsonMarshalRecursive(...)
	pkg/util/request_tracker/request_extractor.go:108 +0x154
github.com/cortexproject/cortex/pkg/util/request_tracker.trimForJsonMarshal(...)
	pkg/util/request_tracker/request_extractor.go:96
github.com/cortexproject/cortex/pkg/util/request_tracker.generateJSONEntryWithTruncatedField(...)
	pkg/util/request_tracker/request_extractor.go:78 +0xa8
github.com/cortexproject/cortex/pkg/util/request_tracker.(*ApiExtractor).Extract(...)
	pkg/util/request_tracker/request_extractor.go:49 +0x254
```
(A control case with valid multi-byte UTF-8 — `strings.Repeat("世", 400)` — does not panic, which is why the existing tests pass.)
</details>

**Expected behavior**
A malformed `match[]`/`query` value should at worst produce a truncated or empty tracker entry; it must never panic, and certainly must not be able to terminate a querier.

**Environment:**
 - Infrastructure: any
 - Deployment tool: any
 - Affected versions: master (85a4553). The unbounded scan predates #7640; that PR is where the panic class was last addressed.

**Additional Context**
Suggested fix — bound the scan:
```go
for size > 0 && !utf8.RuneStart(bytesStr[size]) {
	size--
}
```
Reaching `size == 0` on a leading continuation byte then yields an empty string, which is the correct degenerate result. It would also be worth adding a `recover()` to the request-handling goroutine in `pkg/querier/worker/scheduler_processor.go` so that no future extractor bug can take down a querier process, and extending the #7640 regression tests with invalid-UTF-8 inputs.

Found during a review of recent master commits.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -86,6 +86,7 @@
 * [BUGFIX] Distributor: Return HTTP 499 (Client Closed Request) instead of 500 when a remote-write or OTLP push is canceled by the client, so client-side cancellations are no longer counted as server-side errors. #7717
 * [BUGFIX] Querier: Fix gRPC `codes.Canceled` errors being mapped to HTTP 500 instead of 499 when a client cancels a query. #7738
 * [BUGFIX] Compactor: Fix spurious `bucket operation fail after retries` error logs emitted during partial block cleanup. #7749
+* [BUGFIX] Querier: Fix panic (`index out of range [-1]`) in the active request tracker when truncating a `match[]`/`query` value made entirely of invalid UTF-8 continuation bytes. The backwards scan for a rune boundary now stops at index 0 instead of underflowing. #7729
 
 ## 1.21.1 2026-06-04
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -91,6 +91,7 @@
 * [BUGFIX] Parquet Converter: Fix `auto_forget_delay` having no effect. The ring lifecycler was created without the auto-forget delegate, so unhealthy instances were never automatically removed from the ring. #7752
 * [BUGFIX] Alertmanager: Reject the global `mattermost_webhook_url_file` setting in per-tenant configs, consistent with every other global `*_file` setting. #7768
 * [BUGFIX] Alertmanager: Tighten per-tenant config validation to reject additional file-based settings. #7767
+* [BUGFIX] Querier: Fix panic (`index out of range [-1]`) in the active request tracker when truncating a `match[]`/`query` value made entirely of invalid UTF-8 continuation bytes. The backwards scan for a rune boundary now stops at index 0 instead of underflowing. #7729
 
 ## 1.21.1 2026-06-04
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -91,7 +91,7 @@
 * [BUGFIX] Parquet Converter: Fix `auto_forget_delay` having no effect. The ring lifecycler was created without the auto-forget delegate, so unhealthy instances were never automatically removed from the ring. #7752
 * [BUGFIX] Alertmanager: Reject the global `mattermost_webhook_url_file` setting in per-tenant configs, consistent with every other global `*_file` setting. #7768
 * [BUGFIX] Alertmanager: Tighten per-tenant config validation to reject additional file-based settings. #7767
-* [BUGFIX] Querier: Fix panic (`index out of range [-1]`) in the active request tracker when truncating a `match[]`/`query` value made entirely of invalid UTF-8 continuation bytes. The backwards scan for a rune boundary now stops at index 0 instead of underflowing. #7729
+* [BUGFIX] Querier: Fix panic (`index out of range [-1]`) in the active request tracker when truncating a `match[]`/`query` value made entirely of invalid UTF-8 continuation bytes. The backwards scan for a rune boundary now stops at index 0 instead of underflowing. #7743
 
 ## 1.21.1 2026-06-04
 

--- a/pkg/util/request_tracker/request_extractor.go
+++ b/pkg/util/request_tracker/request_extractor.go
@@ -83,7 +83,7 @@ func trimStringByBytes(str string, size int) string {
 	bytesStr := []byte(str)
 	trimIndex := len(bytesStr)
 	if size < len(bytesStr) {
-		for !utf8.RuneStart(bytesStr[size]) {
+		for size > 0 && !utf8.RuneStart(bytesStr[size]) {
 			size--
 		}
 		trimIndex = size

--- a/pkg/util/request_tracker/request_tracker_test.go
+++ b/pkg/util/request_tracker/request_tracker_test.go
@@ -161,6 +161,20 @@ func TestTrimForJsonMarshalMultiByteUTF8(t *testing.T) {
 	}
 }
 
+// TestTrimForJsonMarshalInvalidUTF8 reproduces a panic where a string made
+// entirely of UTF-8 continuation bytes (0x80-0xBF, no valid rune-start byte)
+// caused the backwards scan for a rune boundary to underflow past index 0
+// and index the byte slice with a negative index.
+func TestTrimForJsonMarshalInvalidUTF8(t *testing.T) {
+	invalid := strings.Repeat("\x80", 1200)
+
+	require.NotPanics(t, func() {
+		out := trimForJsonMarshal(invalid, 800)
+		assert.True(t, utf8.ValidString(out), "result should be valid UTF-8")
+		assert.Equal(t, "", out)
+	})
+}
+
 // TestGenerateJSONEntryWithTruncatedFieldNegativeSize reproduces the request
 // tracker panic where a multi-byte UTF-8 field had to be truncated to a
 // negative remaining size because the rest of the entry already consumed the

